hs_composite_resource/models.py

```diff
@@ -127,6 +212,145 @@ def can_set_folder_to_fileset(self, dir_path):
+        otherwise, return True
+        """
+        return self._can_set_folder_to_mi_or_mp_aggregation(dir_path=dir_path, aggr_type="ModelProgram")
+
+    def _can_set_folder_to_mi_or_mp_aggregation(self, dir_path, aggr_type):
+        """helper to check if the specified folder *dir_path* can be set to ModelProgram or ModelInstance aggregation
+
+        :param dir_path: Resource file directory path (full folder path starting with resource id)
+        for which the aggregation of the type *aggr_type* to be set
+        :param aggr_type: a value of either ModelProgram or ModelInstance
+
+        :return True or False
+
+        """
+
+        # checking target folder for any aggregation
+        if self.get_folder_aggregation_object(dir_path) is not None:
+            # target folder is already an aggregation
+            return False
+
+        # checking sub-folders for fileset aggregation
+        # check that we don't have any sub folder of dir_path representing a fileset aggregation
+        # so that we can avoid nesting a fileset aggregation inside a model program or model instance aggregation
+        if self.filesetlogicalfile_set.filter(folder__startswith=dir_path).exists():
+            return False
+
+        if aggr_type == "ModelProgram":
+            # checking sub-folders for model program aggregation
+            # check that we don't have any sub folder of dir_path representing a model program aggregation
+            # so that we can avoid nesting a model program aggregation inside a model
+            # program aggregation
+            if self.modelprogramlogicalfile_set.filter(folder__startswith=dir_path).exists():
+                return False
+
+        # checking sub-folders for model instance aggregation
+        # check that we don't have any sub folder of dir_path representing a model instance aggregation
+        # so that we can avoid nesting a model instance aggregation inside a model program aggregation
+        if self.modelinstancelogicalfile_set.filter(folder__startswith=dir_path).exists():
+            return False
+
+        # check the first parent folder that represents an aggregation
+        irods_path = dir_path
+        if self.is_federated:
+            irods_path = os.path.join(self.resource_federation_path, irods_path)
+
+        # get the parent folder path
+        path = os.path.dirname(dir_path)
+        parent_aggregation = None
+        while '/' in path:
+            if path == self.file_path:
+                break
+            parent_aggregation = self.get_folder_aggregation_object(path)
+            if parent_aggregation is not None:
+                # this is the first parent folder that represents an aggregation
+                break
+            # get the next parent folder path
+            path = os.path.dirname(path)
+
+        if parent_aggregation is not None:
+            if aggr_type == "ModelProgram":
```

Comment:
ModelProgram agg type is checked for 3 times in this method.  I'm having trouble following the logic of this method.